sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableServiceImplTest.java

```diff
@@ -126,6 +143,538 @@ public void testRead() throws IOException {
+
+    when(mockBigtableDataClient.readFlatRows(any(ReadRowsRequest.class), any()))
+        .thenAnswer(new Answer<ScanHandler>() {
+          @Override
+          public ScanHandler answer(InvocationOnMock invocationOnMock) throws Throwable {
+            StreamObserver<FlatRow> flatRowObserver = invocationOnMock.getArgument(1);
+            new Thread() {
+              @Override
+              public void run() {
+                for (int i = 0; i < MINI_BATCH_ROW_LIMIT; i++) {
+                  flatRowObserver.onNext(expectedFirstRangeRows.get(i));
+                }
+                flatRowObserver.onCompleted();
+              }
+            }.start();
+
+            return scanHandler;
+          }
+        }).thenAnswer(new Answer<ScanHandler>() {
+          @Override
+          public ScanHandler answer(InvocationOnMock invocationOnMock) throws Throwable {
+            StreamObserver<FlatRow> flatRowObserver = invocationOnMock.getArgument(1);
+            new Thread() {
+              @Override
+              public void run() {
+                for (int i = 0; i < MINI_BATCH_ROW_LIMIT; i++) {
+                  flatRowObserver.onNext(expectedSecondRangeRows.get(i));
+                }
+                flatRowObserver.onCompleted();
+              }
+            }.start();
+
+            return scanHandler;
+          }
+        }).thenAnswer(new Answer<ScanHandler>() {
+          @Override
+          public ScanHandler answer(InvocationOnMock invocationOnMock) throws Throwable {
+            StreamObserver<FlatRow> flatRowObserver = invocationOnMock.getArgument(1);
+            flatRowObserver.onCompleted();
+            return scanHandler;
+          }
+        });
+
+    BigtableService.Reader underTest =
+        new BigtableServiceImpl.BigtableMiniBatchReaderImpl(mockSession, mockBigtableSource);
+
+    underTest.start();
+    Assert.assertEquals(FlatRowConverter.convert(expectedFirstRangeRows.get(0)), underTest.getCurrentRow());
+    for (int i = 1; i < MINI_BATCH_ROW_LIMIT; i++) {
+      underTest.advance();
+    }
+    for (int i = 0; i < MINI_BATCH_ROW_LIMIT; i++) {
+      underTest.advance();
+    }
+    Assert.assertEquals(
+        FlatRowConverter.convert(expectedSecondRangeRows.get(expectedSecondRangeRows.size() - 1)), underTest.getCurrentRow());
+    Assert.assertFalse(underTest.advance());
+
+    underTest.close();
+    verifyMetricWasSet("google.bigtable.v2.ReadRows", "ok", 2);
+  }
+
+  /**
+   * This test ensures that all the rows are properly added to the buffer and read. This example
+   * uses two ranges with MINI_BATCH_ROW_LIMIT rows. The buffer should be refilled twice and
+   * ReadRowsAsync should be called twice. The following test follows this example: FirstRange:
+   * [a,b1,...,b99,c) SecondRange: [c,d1,...,d99,e)
+   *
+   * @throws IOException
+   */
+  @Test
+  public void testReadMultipleRanges() throws IOException {
+    List<FlatRow> expectedFirstRangeRows = new ArrayList<>();
```

Comment:
consider extracting a helper function that makes rows for you: List<FlatRow> createRow(String prefix, int count)

Also, your current logic creates an invalid ordering of rows.
Current key order: a, a1,...a5,...a20
But a20 sorts before a5. You can easily fix this by adding leading zero format specifier